### PR TITLE
Upgrade Spock2 to milestone 5

### PR DESCRIPTION
**Pull request by rieske**
Contains fixes for known issues in our tests:
- https://github.com/spockframework/spock/issues/1267 - incorrect reporting of failed tests when using `@Retry` with `SETUP_FEATURE_CLEANUP` mode
- https://github.com/spockframework/spock/issues/1263 - jUnit assumption handling from superclasses - we had a workaround implemented for this which goes away now.

Release notes: https://spockframework.org/spock/docs/2.0-M5/release_notes.html#_2_0_m5_2021_03_23

### CODE REVIEW

**Comment by rieske** [comment]
This change required adjusting of some visibility modifiers in tests to overcome new compilation failures like the following: https://ge.gradle.org/s/tapccphi43va2/console-log#L2120

It looks like the new version does not like when a Groovy test class has an explicit getter when a field of that name exists: https://ge.gradle.org/s/3gwxqi5grqppq/console-log?anchor=1511
There were also compilation problems where a base class would declare an abstract getter that would be implemented using a Groovy field with a final modifier in a subclass. Removing offending modifiers solved those problems.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ParallelDownloadsIntegrationTest.groovy [hunk 1]
```diff
@@ -24,10 +24,10 @@ import spock.lang.Unroll
 
 class ParallelDownloadsIntegrationTest extends AbstractHttpDependencyResolutionTest {
     @Rule
-    BlockingHttpServer server = new BlockingHttpServer()
+    public BlockingHttpServer blockingServer = new BlockingHttpServer()
```
Comment (rieske) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ParallelDownloadsIntegrationTest.groovy:27]: This field started clashing with a `server` field declared in a superclass.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultNamedDomainObjectCollectionTest.groovy [hunk 2]
```diff
@@ -31,15 +31,15 @@ class DefaultNamedDomainObjectCollectionTest extends AbstractNamedDomainObjectCo
...
     Instantiator instantiator = TestUtil.instantiatorFactory().decorateLenient()
 
-    final DefaultNamedDomainObjectCollection<Bean> container = new DefaultNamedDomainObjectCollection<Bean>(Bean, new IterationOrderRetainingSetElementSource<Bean>(), instantiator, namer, callbackActionDecorator)
-    final Bean a = new BeanSub1("a")
-    final Bean b = new BeanSub1("b")
-    final Bean c = new BeanSub1("c")
-    final Bean d = new BeanSub2("d")
-    final boolean externalProviderAllowed = true
+    DefaultNamedDomainObjectCollection<Bean> container = new DefaultNamedDomainObjectCollection<Bean>(Bean, new IterationOrderRetainingSetElementSource<Bean>(), instantiator, namer, callbackActionDecorator)
+    Bean a = new BeanSub1("a")
```
Comment (bmuskalla) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultNamedDomainObjectCollectionTest.groovy:35]: Do they have any intention of fixing this? This sounds like a major breakage (esp given the ripple effect it has on buildscript authors)

**Review by bmuskalla — COMMENTED** [review]
I'm a bit worried about the `final` changes as this will cause a churn in build script tests again. Do you know if they intend to fix this? I wonder if we should hold this off until they do

File: subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultNamedDomainObjectCollectionTest.groovy (hunk 2 above)
Comment (bmuskalla) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultNamedDomainObjectCollectionTest.groovy:35]: Ignore this, didn't realize these are methods :(

**Review by bmuskalla — APPROVED** [review]

**Comment by rieske** [comment]
@bot-gradle test and merge this

File: subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultNamedDomainObjectCollectionTest.groovy (hunk 2 above)
Comment (rieske) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultNamedDomainObjectCollectionTest.groovy:35]: I am now trying to come up with a minimal reproducer for this issue and I am getting failures for the same cases even with m4 which in our project did work.
For example:
```
class GetterClash extends spock.lang.Specification {
    final String bar = "BAR"
    String getBar() {
        return bar
    }
    def "some test"() {
        expect:
        true
    }
}
```
Fails with m4 (at test execution time):
```
java.lang.IllegalAccessError: Update to non-static final field GetterClash.bar attempted from a different method ($spock_initializeFields) than the initializer method <init> 
	at GetterClash.$spock_initializeFields(GetterClash.groovy)
```
Fails with m5 (at compile time and with a better message):
```
> Task :compileTestGroovy FAILED
startup failed:
/home/vaidotas/tmp/gradle/spock/m5/src/test/groovy/GetterClash.groovy: 6: Final field 'bar' conflicts with method 'getBar'; please rename either of them @ line 6, column 5.
       final String bar = "BAR"
```
In this case, removing the getter solves the "problem" (I did this in `AbstractIntegrationSpec` for `mavenRepo/ivyRepo` fields.


Same story with:
```
abstract class BaseSpec extends spock.lang.Specification {
    abstract String getFoo()
    def "base test"() {
        expect:
        true
    }
}
```
and
```
class ChildSpec extends BaseSpec {
    // this is effectively @Override int getFoo() { return 42 }
    // either being explicit that this is an override or removing final modifier solves the problem
    final String foo = "FOO";
    def "child test"() {
        expect:
        true
    }
}
```
With m4, I am seeing test failures:
```
java.lang.IllegalAccessError: Update to non-static final field ChildSpec.foo attempted from a different method ($spock_initializeFields) than the initializer method <init> 
	at ChildSpec.$spock_initializeFields(ChildSpec.groovy)
```
While m5 fails me at compile time:
```
> Task :compileTestGroovy FAILED
startup failed:
/home/vaidotas/tmp/gradle/spock/m5/src/test/groovy/ChildSpec.groovy: 3: Final field 'foo' conflicts with method 'getFoo'; please rename either of them @ line 3, column 5.
       final String foo = "FOO";
```

I do not understand why this worked with m4 in our case.

**Comment by rieske** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
This one is being built: https://builds.gradle.org/viewQueued.html?itemId=42321457&tab=queuedBuildOverviewTab

**Comment by rieske** [comment]
@bot-gradle test and merge

**Comment by rieske** [comment]
@bot-gradle test and merge

**Pull request merged**